src/platform_impl/web/event_loop/proxy.rs

```diff
@@ -1,26 +1,99 @@
-use super::runner;
-use crate::event::Event;
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::mpsc::{self, Receiver, RecvError, SendError, Sender, TryRecvError};
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll, Waker};
+
 use crate::event_loop::EventLoopClosed;
 
 pub struct EventLoopProxy<T: 'static> {
-    runner: runner::Shared<T>,
+    sender: AsyncSender<T>,
 }
 
 impl<T: 'static> EventLoopProxy<T> {
-    pub fn new(runner: runner::Shared<T>) -> Self {
-        Self { runner }
+    pub fn new(sender: AsyncSender<T>) -> Self {
+        Self { sender }
     }
 
     pub fn send_event(&self, event: T) -> Result<(), EventLoopClosed<T>> {
-        self.runner.send_event(Event::UserEvent(event));
-        Ok(())
+        match self.sender.send(event) {
+            Ok(()) => Ok(()),
+            Err(SendError(val)) => Err(EventLoopClosed(val)),
+        }
     }
 }
 
 impl<T: 'static> Clone for EventLoopProxy<T> {
     fn clone(&self) -> Self {
         Self {
-            runner: self.runner.clone(),
+            sender: self.sender.clone(),
+        }
+    }
+}
+
+pub fn channel<T: 'static>() -> (AsyncSender<T>, AsyncReceiver<T>) {
+    let (sender, receiver) = mpsc::channel();
+    let waker = Arc::new(Mutex::new(None));
+
+    let sender = AsyncSender {
+        sender,
+        waker: Arc::clone(&waker),
+    };
+    let receiver = AsyncReceiver { receiver, waker };
+
+    (sender, receiver)
+}
+
+pub struct AsyncSender<T: 'static> {
+    sender: Sender<T>,
+    waker: Arc<Mutex<Option<Waker>>>,
+}
+
+impl<T: 'static> AsyncSender<T> {
+    pub fn send(&self, event: T) -> Result<(), SendError<T>> {
+        self.sender.send(event)?;
+
+        if let Some(waker) = self.waker.lock().unwrap().take() {
+            waker.wake();
+        }
+
+        Ok(())
+    }
+}
+
+impl<T: 'static> Clone for AsyncSender<T> {
+    fn clone(&self) -> Self {
+        Self {
+            sender: self.sender.clone(),
+            waker: self.waker.clone(),
+        }
+    }
+}
+
+pub struct AsyncReceiver<T: 'static> {
+    receiver: Receiver<T>,
+    waker: Arc<Mutex<Option<Waker>>>,
+}
+
+impl<T: 'static> Future for AsyncReceiver<T> {
+    type Output = Result<T, RecvError>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        match self.receiver.try_recv() {
+            Ok(event) => Poll::Ready(Ok(event)),
+            Err(TryRecvError::Empty) => {
+                *self.waker.lock().unwrap() = Some(cx.waker().clone());
+
+                match self.receiver.try_recv() {
+                    Ok(event) => {
+                        self.waker.lock().unwrap().take();
+                        Poll::Ready(Ok(event))
+                    }
+                    Err(TryRecvError::Empty) => Poll::Pending,
+                    Err(TryRecvError::Disconnected) => Poll::Ready(Err(RecvError)),
```

Comment:
I take out the `Waker` again if we aren't waiting anymore, which is great for optimization, but on the other hand we are using a `Mutex` which kinda makes this worse?

I don't take out the `Waker` if we fail, because our listener aborts in that case anyway.

Reply:
I made a small benchmark with this, turns out it's pretty awful, so I just removed it.